File: nn_model.jl
using Flux
using Flux: @epochs
using Flux.Optimise: ADAM, train!
using Flux.Data: DataLoader
using TextAnalysis
using Random
using CSV

@info "Loading data..."
df = CSV.read("data/imdb_dataset.csv")
# df = df[1:1000, :]

function preprocess(doc::String)
    doc = StringDocument(doc)
    remove_case!(doc)
    prepare!(doc, strip_html_tags | strip_punctuation | strip_non_letters | strip_stopwords)
    stem!(doc)
    return doc
end


@info "Data preprocessing..."
as_binary(x) = Int(x == "positive")
df.sentiment = map(as_binary, df.sentiment);

doc_list = map(preprocess, df.review);
corpus = Corpus(doc_list);

update_lexicon!(corpus);
m = DocumentTermMatrix(corpus);

tfdf = Matrix(tf_idf(m));

samples = size(tfdf, 1)
perm = randperm(samples);
train_perc = 0.75
train_count = Int(floor(train_perc*samples))

train_x = tfdf[perm[1:train_count], :]
test_x = tfdf[perm[train_count:end], :]
train_y = df.sentiment[perm[1:train_count]]
test_y = df.sentiment[perm[train_count:end]]

# data = DataLoader(transpose(tf), hcat(df.sentiment...), shuffle=true, batchsize=64)
train_data = DataLoader(transpose(train_x), hcat(train_y...), shuffle=true, batchsize=64)
test_data = DataLoader(transpose(test_x), hcat(test_y...), shuffle=true, batchsize=64)

input_size = size(tfdf, 2)
epochs = 50

model = Chain(
    Dense(input_size, 100, relu),
    Dropout(0.5),
    Dense(100, 100, relu),
    Dropout(0.5),
    Dense(100, 1),
)

function loss_all(data, model)
    l = 0f0
    for (x, y) in data
        l += Flux.mse(model(x), y)
    end
    l / length(data)
end

@info "Training model..."
loss(x,y) = Flux.mse(model(x), y)
opt = ADAM()

train_loss = []
test_loss = []
for e in 1:epochs
    @info "EPOCH: $e"
    train!(loss, Flux.params(model), train_data, opt)
    push!(train_loss, loss_all(train_data, model))
    push!(test_loss, loss_all(test_data, model))
end

println(train_loss)
println(test_loss)
# println("Train Loss: ")
# for i in train_loss
#     println(i)
# end
# println("Test Loss: ")
# for i in test_loss
#     println(i)
# end

File: nn_model.py
import re
import pandas as pd
from numpy import vstack
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import (
    f1_score,
    accuracy_score,
    confusion_matrix,
    roc_auc_score,
    balanced_accuracy_score,
)

from keras.models import Sequential
from keras.layers import Dense, Dropout
from keras.callbacks import EarlyStopping, ModelCheckpoint

import matplotlib.pyplot as plt


wnl = WordNetLemmatizer()


def clean(text):
    text = re.sub("<br\s/>", "", text)
    text = text.lower()
    text = re.sub("[^a-z ]", " ", text)
    text = re.sub("\s+", " ", text)
    out = []
    for word in word_tokenize(text):
        out.append(wnl.lemmatize(word))
    return " ".join(out)


def load_clean_data():
    df = pd.read_csv("data/imdb_dataset.csv")
    df["review"] = df.review.apply(clean)
    df["sentiment"] = [1 if x == "positive" else 0 for x in df.sentiment]
    return df


def create_model(input_size, layer_size, dropout_rate=0.5):
    model = Sequential(
        [
            Dense(
                layer_size,
                input_shape=(input_size,),
                activation="relu",
                kernel_initializer="uniform",
            ),
            Dropout(dropout_rate),
            Dense(1, activation="sigmoid"),
        ]
    )
    model.compile(optimizer="Adam", loss="binary_crossentropy", metrics=["accuracy"])
    return model


def main():
    tf_vec = TfidfVectorizer(
        strip_accents="ascii", lowercase=True, stop_words="english"
    )
    df = load_clean_data()
    X = tf_vec.fit_transform(df.review)
    y = df.sentiment.values

    train_x, test_x, train_y, test_y = train_test_split(
        X, y, train_size=0.75, stratify=y
    )

    model = create_model(X.shape[1], 5)

    epochs = 100

    callbacks = [
        EarlyStopping(patience=5, monitor="val_loss", mode="min", verbose=1),
        ModelCheckpoint(
            filepath="models/model.{epoch:02d}-{val_loss:.3f}.h5",
            monitor="val_loss",
            mode="min",
            save_best_only=True,
            verbose=1,
        ),
    ]

    history = model.fit(
        train_x,
        vstack(train_y),
        validation_data=(test_x, test_y),
        epochs=epochs,
        batch_size=32,
        callbacks=callbacks,
    )

    pred = model.predict_classes(test_x)

    print("===============Metrics===============")
    print(f"## F1 Score: {f1_score(test_y, pred)}")
    print(f"## Accuracy: {accuracy_score(test_y, pred)}")
    print(f"## Confusion Matrix: \n{confusion_matrix(test_y, pred)}")
    print(f"## Balanced Accuracy: {balanced_accuracy_score(test_y, pred)}")
    print(f"## ROC AUC Score: {roc_auc_score(test_y, pred)}")
    print("=====================================")

    plt.figure(figsize=((15, 10)))
    e = len(history.history["loss"])
    plt.title("Model Loss")
    plt.plot(range(1, e + 1), history.history["loss"], "b.")
    plt.plot(range(1, e + 1), history.history["val_loss"], "r")
    plt.axvline(x=3)

    plt.xlabel("Epochs")
    plt.ylabel("Loss")
    plt.legend(labels=["Train Loss", "Val Loss"])
    plt.show()


if __name__ == "__main__":
    main()
